File: webapp/src/lib/money.ts
const E: Record<string, number> = {
  USD: 2, EUR: 2, GBP: 2, AED: 2, RUB: 2, CHF: 2, CAD: 2, AUD: 2,
  CNY: 2, INR: 2, TRY: 2, PLN: 2, CZK: 2, SEK: 2,
  JPY: 0, KRW: 0,
  KWD: 3, BHD: 3, OMR: 3, TND: 3,
};

export const exp = (c: string): number => E[c.toUpperCase()] ?? 2;

export const fromMinor = (m: number, c: string): number => m / 10 ** exp(c);

export const toMinor = (n: number, c: string): number => Math.round(n * 10 ** exp(c));

export function formatAmount(m: number, c: string): string {
  const e = exp(c);
  const sign = m < 0 ? "−" : "";
  const n = Math.abs(m);
  if (e === 0) return `${sign}${n} ${c}`;
  const w = Math.floor(n / 10 ** e);
  const f = (n % 10 ** e).toString().padStart(e, "0");
  return `${sign}${w}.${f} ${c}`;
}

/** USD-style formatter for base currency: $1,234.50 */
export function formatBase(m: number, c: string): string {
  const e = exp(c);
  const sign = m < 0 ? "−" : "";
  const n = Math.abs(m);
  if (e === 0) return `${sign}${symbolFor(c)}${n.toLocaleString("en-US")}`;
  const whole = Math.floor(n / 10 ** e).toLocaleString("en-US");
  const frac = (n % 10 ** e).toString().padStart(e, "0");
  return `${sign}${symbolFor(c)}${whole}.${frac}`;
}

function symbolFor(c: string): string {
  switch (c.toUpperCase()) {
    case "USD": case "AUD": case "CAD": return "$";
    case "EUR": return "€";
    case "GBP": return "£";
    case "RUB": return "₽";
    case "JPY": case "CNY": return "¥";
    default: return "";
  }
}

File: src/finance_app/domain/money.py
from decimal import ROUND_HALF_EVEN, Decimal

# Minimal table; extend as needed.
_EXPONENTS = {
    "USD": 2, "EUR": 2, "GBP": 2, "AED": 2, "RUB": 2, "CHF": 2, "CAD": 2,
    "AUD": 2, "CNY": 2, "INR": 2, "TRY": 2, "PLN": 2, "CZK": 2, "SEK": 2,
    "JPY": 0, "KRW": 0,
    "KWD": 3, "BHD": 3, "OMR": 3, "TND": 3,
}

def exponent(code: str) -> int:
    return _EXPONENTS.get(code.upper(), 2)

def to_minor(amount: float | int | Decimal | str, code: str) -> int:
    q = Decimal(10) ** -exponent(code)
    d = Decimal(str(amount)).quantize(q, rounding=ROUND_HALF_EVEN)
    return int(d * (Decimal(10) ** exponent(code)))

def from_minor(minor: int, code: str) -> float:
    return float(Decimal(minor) / (Decimal(10) ** exponent(code)))

def format_amount(minor: int, code: str) -> str:
    e = exponent(code)
    sign = "-" if minor < 0 else ""
    n = abs(minor)
    if e == 0:
        return f"{sign}{n} {code}"
    whole, frac = divmod(n, 10**e)
    return f"{sign}{whole}.{frac:0{e}d} {code}"

def convert(minor_src: int, src: str, dst: str, *, rate: float) -> int:
    """Convert `minor_src` (in src minor units) to dst minor units using rate src→dst."""
    if src.upper() == dst.upper():
        return minor_src
    src_e = exponent(src)
    dst_e = exponent(dst)
    # scale: minor_dst = minor_src * rate * 10^(dst_e - src_e)
    factor = Decimal(rate) * (Decimal(10) ** (dst_e - src_e))
    return int((Decimal(minor_src) * factor).quantize(Decimal(1), rounding=ROUND_HALF_EVEN))
